iso_nso_calculator/index.ts

```diff
@@ -0,0 +1,221 @@
+function addISOColumns(vestingData: VestingScheduleWithYearAndGrantId[], isoData: VestingDataWithShares[]): VestingDataWithISO[] {
+  const result: VestingDataWithISO[] = [];
+  const isoRemainingByGrantAndYear: { [key: string]: number } = {};
+
+  for (let i = 0; i < vestingData.length; i++) {
+    // Get the corresponding ISO share information for the grant and year
+    const isoInfo = isoData.find((iso) => iso.Grant === vestingData[i].Grant && iso.Year === vestingData[i].Year);
+
+    if (!isoInfo) {
+      throw new Error(`ISO data not found for Grant ${vestingData[i].Grant} and Year ${vestingData[i].Year}`);
+    }
+
+    const grantYearKey = `${vestingData[i].Grant}-${vestingData[i].Year}`;
+
+    // Initialize ISO Remaining at the start of the year
+    if (!(grantYearKey in isoRemainingByGrantAndYear)) {
+      isoRemainingByGrantAndYear[grantYearKey] = isoInfo.ISOShares;
+    }
+
+    // Calculate ISO Used as the minimum of Amount Vested and ISO Remaining
+    const isoUsed = Math.min(vestingData[i]["Event Quantity"], isoRemainingByGrantAndYear[grantYearKey]);
+
+    // Calculate NSO if ISO Remaining is zero
+    const nso = isoRemainingByGrantAndYear[grantYearKey] === 0 ? vestingData[i]["Event Quantity"] : Math.max(0, vestingData[i]["Event Quantity"] - isoUsed);
+
+    // Update ISO Remaining
+    const isoRemaining = isoRemainingByGrantAndYear[grantYearKey] - isoUsed;
+
+    // Store the updated ISO Remaining for the next rows in the same year
+    isoRemainingByGrantAndYear[grantYearKey] = isoRemaining;
+
+    // Add the row with the new ISO and NSO columns
+    result.push({
+      ...vestingData[i],
+      ISO: isoUsed,
+      NSO: nso,
+      ISORemaining: isoRemaining,
+    });
+  }
+
+  return result;
+}
+
+export const isoNsoCalculator = (packagePath: string, stakeholderId: string, capacity: number): any => {
+  const ocfPackage: OcfPackageContent = readOcfPackage(packagePath);
+
+  const valuations = ocfPackage.valuations;
+  const transactions = ocfPackage.transactions;
+  const equityCompensationIssuances: any[] = transactions.filter((transaction: any) => transaction.stakeholder_id === stakeholderId && transaction.object_type === "TX_EQUITY_COMPENSATION_ISSUANCE");
+
+  if (equityCompensationIssuances.length === 0) {
+    throw new Error("No equity compensation issuances found for stakeholder");
+  }
+
+  const sortedIssuances = equityCompensationIssuances.sort((a: { date: string }, b: { date: string }) => a.date.localeCompare(b.date));
+
+  const combinedYearTable: any[] = [];
+  const combinedGrants: any[] = [];
+  let vestedByYearTable: any[] = [];
+
+  sortedIssuances.forEach((issuance: any) => {
+    const vestingSchedule = generateSchedule(packagePath, issuance.security_id);
+    const vestingScheduleWithYearAndGrantId = addYearAndGrantId(vestingSchedule, issuance.id);
+    combinedGrants.push(vestingScheduleWithYearAndGrantId);
+    vestedByYearTable = sumByYear(vestingScheduleWithYearAndGrantId);
```

Comment:
instead of summing by year and then applying the ISO test to the aggregate FMV for that year, we should apply the test on an installment by installment basis.